src/erc7496/erc7496.cairo

```diff
@@ -0,0 +1,155 @@
+        // @param traitKey The trait key to get the value of
+        fn _get_trait_value(
+            self: @ComponentState<TContractState>, token_id: u256, trait_key: felt252
+        ) -> felt252 {
+            // Return the trait value.
+            self.ERC7496_traits.read((token_id, trait_key))
+        }
+
+        // @notice Get the values of traits for a given token ID.
+        // @param tokenId The token ID to get the trait values for
+        // @param traitKeys The trait keys to get the values of
+        fn _get_trait_values(
+            self: @ComponentState<TContractState>, token_id: u256, trait_keys: Span<felt252>
+        ) -> Span<felt252> {
+            let mut trait_values = array![];
+            // Assign each trait value to the corresponding key.
+            let mut index = 0;
+            while index < trait_keys
+                .len() {
+                    trait_values
+                        .append(self.ERC7496_traits.read((token_id, *trait_keys.at(index))));
+                    index += 1;
+                };
+            trait_values.span()
+        }
+
+        // @notice Set the value of a trait for a given token ID.
+        //         Reverts if the trait value is unchanged.
+        // @dev    IMPORTANT: Override this method with access role restriction.
+        // @param tokenId The token ID to set the trait value for
+        // @param traitKey The trait key to set the value of
+        // @param newValue The new trait value to set
+        fn _set_trait(
+            ref self: ComponentState<TContractState>,
+            token_id: u256,
+            trait_key: felt252,
+            new_value: felt252
+        ) {
+            // Revert if the new value is the same as the existing value.
+            let existing_value = self.ERC7496_traits.read((token_id, trait_key));
+            assert(existing_value != new_value, Errors::TRAIT_VALUE_UNCHANGED);
+            // Set the new trait value.
+            self._set_trait_internal(token_id, trait_key, new_value);
+            // Emit the event noting the update.
+            self.emit(TraitUpdated { trait_key, token_id, trait_value: new_value });
+        }
+
+        // @notice Set the trait value (without emitting an event).
+        // @param tokenId The token ID to set the trait value for
+        // @param traitKey The trait key to set the value of
+        // @param newValue The new trait value to set
+        fn _set_trait_internal(
+            ref self: ComponentState<TContractState>,
+            token_id: u256,
+            trait_key: felt252,
+            new_value: felt252
+        ) {
+            // Set the new trait value.
+            self.ERC7496_traits.write((token_id, trait_key), new_value);
+        }
+
+        // @notice Set the URI for the trait metadata.
+        // @param uri The new URI to set.
+        fn _set_trait_metadata_uri(ref self: ComponentState<TContractState>, uri: ByteArray) {
```

Comment:
agree to make it private as the IERC7496 doesn't expose any function that can modify the metadata uri